### ISSUE TITLE
bad error locations reported when using  CompilationMessage.span to format error messages

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#31652]
When I try compiling Flutter using front-end I have all errors reported on line 1 column X.

```
[ +160 ms] compile debug message: Warning: line 1, column 10606 of /usr/local/google/home/vegorov/src/flutter-clean/flutter/packages/flutter/lib/src/rendering/binding.dart: Superclass has no method named 'hitTest'.
```

I traced it down to `CompilationMessage`-s that are passed to `onError` listeners. Apparently they either don't have correct source location information set or we are not using that information correctly. You can repro it by simply running `pkg/vm/tool/gen_kernel` on anything that has a compilation error, this would produce two errors, one of which comes from `onError` listener and one from somewhere else (Fasta itself?):

```
╭─vegorov@volhv ~/src/dart/sdk ‹fix-socket U›
╰─$ pkg/vm/tool/gen_kernel --platform out/ReleaseX64/vm_platform_strong.dill -o t/flu.dart.dill t/flu.dart
Error: line 1, column 176 of t/flu.dart: The getter 'field' isn't defined for the class 'dart.core::String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'field'.
t/flu.dart:16:18: Error: The getter 'field' isn't defined for the class 'dart.core::String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'field'.
  get field => f.field;
```

The first message is formatted using `message.span.message(message)`. If I print the `message.span` then it looks like this:

```
span = <SourceSpanBase: from <SourceLocation: 175 file:///usr/local/google/home/vegorov/src/dart/sdk/t/flu.dart:1:176> to <SourceLocation: 175 file:///usr/local/google/home/vegorov/src/dart/sdk/t/flu.dart:1:176> ""> / SourceSpanBase
```

Peter & Aske, did you ever encounter anything like this? 

If this is an API misuse (e.g. we are not suppose to format `CompilationMessage` using `message.span.message(message)`), please advise what kind of API we should use and reassign to me.

/cc @alexmarkov @jensjoha

**Comment by peter-ahe-google** [comment]
I think there's a bug in `_CompilationMessage` in `pkg/front_end/lib/src/base/processed_options.dart`. It doesn't pass line and column information to SourceLocation which has some dubious semantics in that case, that is, use offset as column.

**Comment by peter-ahe-google** [comment]
I would prefer if you relied on Fasta's built-in error reporting. This is actually tested.

**Comment by a-siva** [comment]
relying the built-in error reporting is not very desirable as we would like to control how messages 
are displayed back to the IDE tool or the command line flutter tool console.
I believe the built-in version just dumps the error messages to stdout.

**Comment by peter-ahe-google** [comment]
TL;DR: Use CompilerOptions.onProblem instead of CompilerOptions.onError and CompilerOptions.reportMessages, but be aware that the API isn't stable.

@a-siva I agree. Once the front-end API has matured, these features should be included and tested. However, for now, we have to avoid CompilationMessage as that API isn't complete or tested.

API maturity is specifically not a goal at this point, and we're instead currently collecting requirements. This particular bug was marked P1, and mentioned nothing about IDE integration. My reply was in that context.

However, this bug report and others got me thinking about this problem and I have since added an API to replace CompilerOptions.onError and CompilerOptions.reportMessages. See commit 51d480385501a1568e0f418838df8e708b9cae0b for more details.

**Comment by peter-ahe-google** [comment]
I believe this issue was fixed by 51d480385501a1568e0f418838df8e708b9cae0b.
